modules/libcom/src/iocsh/atInit.c

```diff
@@ -0,0 +1,101 @@
+/*************************************************************************\
+* Copyright (C) 2020 Dirk Zimoch
+* Copyright (C) 2020-2025 European Spallation Source, ERIC
+* SPDX-License-Identifier: EPICS
+* EPICS BASE is distributed subject to a Software License Agreement found
+* in file LICENSE that is included with this distribution.
+\*************************************************************************/
+
+#include <cantProceed.h>
+#include <ellLib.h>
+#include <errlog.h>
+#include <initHooks.h>
+#include <iocsh.h>
+#include <string.h>
+
+#include "atInit.h"
+
+static const char helpMessage[] =
+    "Allows you to define commands to be run after the iocInit\n"
+    "Example commands:\n"
+    "  atInit \"dbpf <PV> <VAL>\"\n"
+    "  atInit \"date\"\n";
+
+struct cmditem {
+    ELLNODE node;
+    char* cmd;
+};
+
+static ELLLIST s_cmdlist = ELLLIST_INIT;
+static int s_initendflag = 0; // Defines the end of the initialization
+
+static void atInitHook(initHookState state)
+{
+    if (state != initHookAfterIocRunning)
+        return;
+
+    struct cmditem* item = NULL;
+
+    while ((item = (struct cmditem*)ellGet(&s_cmdlist))) {
+        printf("%s\n", item->cmd);
+
+        if (iocshCmd(item->cmd))
+            printf(ERL_ERROR " atInit: "
+                             "command '%s' failed to run\n",
+                   item->cmd);
+
+        free(item);
+    }
+
+    s_initendflag = 1;
+}
+
+static struct cmditem* newItem(char* cmd)
+{
+    size_t cmd_len = strlen(cmd) + 1;
+
+    struct cmditem* item = mallocMustSucceed(sizeof(struct cmditem) + cmd_len,
+                                             ERL_ERROR " atInit: "
+                                                       "failed to allocate memory for cmditem\n");
+    item->cmd = (char*)(item + 1);
+    strncpy(item->cmd, cmd, cmd_len);
```

Comment:
I think this present code is fine.

To be optimally optimal, since the sizes are all known after `strlen()`, `memcpy()` is best so long as the trailing nil is accounted for.

```c
size_t cmd_len = strlen(cmd) + 1;
struct cmditem* item = mallocMustSucceed(sizeof(struct cmditem) + cmd_len, "atInitHook");
item->cmd = (char*)(item + 1);
memcpy(item->cmd, cmd, cmd_len);
```

Also, no error formatting should be used with `mallocMustSucceed()`, the second argument is meant to be just an identifier.  Inserting `ERL_ERROR` into the message printed from `mallocMustSucceed()` would be reasonable, but a separate change.

https://github.com/epics-base/epics-base/blob/144f9756eac4e800f18d3d9a35f0d42a712614fd/modules/libcom/src/misc/cantProceed.c#L43-L47

Reply:
> Inserting `ERL_ERROR` into the message printed from `mallocMustSucceed()`

1d19ba4cc2e8c9ab17288b684022a0682e398bde, which also updates the doc commends for `*MustSucceed()`.